### PR TITLE
Feature/FindNeighbors

### PR DESCRIPTION
**Pull request by mmarineBlueQuartz**
* Added filter
* Added unit test
* Added documentation

### CODE REVIEW

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/FindNeighbors.cpp [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+#include "complex/Parameters/ArraySelectionParameter.hpp"
+#include "complex/Parameters/BoolParameter.hpp"
+#include "complex/Parameters/GeometrySelectionParameter.hpp"
+#include "complex/Parameters/MultiArraySelectionParameter.hpp"
+
+namespace complex
+{
+namespace
+{
+} // namespace
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/FindNeighbors.cpp:20]:
```suggestion
```

File: src/Plugins/ComplexCore/test/FindNeighborsTest.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+namespace
+{
+inline const std::string k_ImageGeomName = "Image Geom";
+inline const std::string k_FeatureIdsName = "Feature IDs";
+inline const std::string k_BoundaryCellsName = "Boundary Cells";
+inline const std::string k_NumNeighborsName = "Num Neighbors";
+inline const std::string k_NeighborListName = "Neighbor List";
+inline const std::string k_SharedSurfaceAreaName = "Shared Surface Area";
+inline const std::string k_SurfaceFeaturesName = "Surface Features";
+inline const std::vector<std::string> k_CellFeatureNames{"Array 1", "Array 2", "Array 3"};
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/test/FindNeighborsTest.cpp:21]:
```suggestion
const std::string k_ImageGeomName = "Image Geom";
const std::string k_FeatureIdsName = "Feature IDs";
const std::string k_BoundaryCellsName = "Boundary Cells";
const std::string k_NumNeighborsName = "Num Neighbors";
const std::string k_NeighborListName = "Neighbor List";
const std::string k_SharedSurfaceAreaName = "Shared Surface Area";
const std::string k_SurfaceFeaturesName = "Surface Features";
const std::vector<std::string> k_CellFeatureNames{"Array 1", "Array 2", "Array 3"};
```

File: src/Plugins/ComplexCore/test/FindNeighborsTest.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  args.insert(FindNeighbors::k_NeighborList_Key, std::make_any<DataPath>(k_NeighborList));
+  args.insert(FindNeighbors::k_SharedSurfaceArea_Key, std::make_any<DataPath>(k_SharedSurfaceArea));
+  args.insert(FindNeighbors::k_SurfaceFeatures_Key, std::make_any<DataPath>(k_SurfaceFeatures));
+
+  auto preflightResult = filter.preflight(ds, args);
+  COMPLEX_RESULT_REQUIRE_VALID(preflightResult.outputActions);
+
+  auto result = filter.execute(ds, args);
+  COMPLEX_RESULT_REQUIRE_VALID(result.result);
+}
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/test/FindNeighborsTest.cpp:93]: Needs a newline at end of file.

File: src/complex/DataStructure/AbstractDataStore.hpp [hunk 4]
```diff
@@ -44,9 +44,13 @@ class AbstractDataStore : public IDataStore
...
 
-    Iterator() = delete;
+    Iterator()
+    : m_DataStore(nullptr)
+    , m_Index(0)
+    {
+    }
 
-    Iterator(AbstractDataStore& dataStore, usize index)
+    Iterator(AbstractDataStore* dataStore, usize index)
```
Comment (JDuffeyBQ) [inline src/complex/DataStructure/AbstractDataStore.hpp:53]:
```suggestion
    Iterator(AbstractDataStore& dataStore, usize index)
```
We can still use a reference in the signature of the constructor even if we store it as a pointer. I also think we should leave a comment that the iterator needs to be default constructible for some standard library algorithm implementations. And this change could be its own PR.

File: src/complex/DataStructure/AbstractDataStore.hpp [hunk 5]
```diff
@@ -166,9 +183,13 @@ class AbstractDataStore : public IDataStore
...
 
-    ConstIterator() = delete;
+    ConstIterator()
+    : m_DataStore(nullptr)
+    , m_Index(0)
+    {
+    }
 
-    ConstIterator(const AbstractDataStore& dataStore, usize index)
+    ConstIterator(const AbstractDataStore* dataStore, usize index)
```
Comment (JDuffeyBQ) [inline src/complex/DataStructure/AbstractDataStore.hpp:192]:
```suggestion
    ConstIterator(const AbstractDataStore& dataStore, usize index)
```
Same as the non-const one.

File: src/complex/Filter/Actions/CreateArrayAction.cpp [hunk 6]
```diff
@@ -55,8 +55,11 @@ Result<> CreateArrayAction::apply(DataStructure& dataStructure, Mode mode) const
   case DataType::float64: {
     return CreateArray<float64>(dataStructure, m_Dims, m_CDims, getCreatedPath(), mode);
   }
+  case DataType::boolean: {
```
Comment (JDuffeyBQ) [inline src/complex/Filter/Actions/CreateArrayAction.cpp:58]: I must missed this in that PR. I can put in quick PR for it.

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/FindNeighbors.cpp [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+    {
+      neighToCount.erase(neighiter);
+    }
+    // Resize the features neighbor list to zero
+    neighborlist[i].resize(0);
+    neighborsurfacearealist[i].resize(0);
+
+    for(const auto [neigh, number] : neighToCount)
+    {
+      float area = float(number) * spacing[0] * spacing[1];
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/FindNeighbors.cpp:317]:
```suggestion
      float area = static_cast<float>(number) * spacing[0] * spacing[1];
```

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/FindNeighbors.cpp [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+  const auto imageGeomNumY = imageGeom.getNumYPoints();
+  const auto imageGeomNumZ = imageGeom.getNumZPoints();
+
+  int64 dims[3] = {
+      static_cast<int64>(udims[0]),
+      static_cast<int64>(udims[1]),
+      static_cast<int64>(udims[2]),
+  };
+
+  int64 neighpoints[6] = {0, 0, 0, 0, 0, 0};
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/FindNeighbors.cpp:172]:
```suggestion
  std::array<int64, 6> neighpoints = {0, 0, 0, 0, 0, 0};
```

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/FindNeighbors.cpp [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+    return MakeErrorResult(-24500, out.str());
+  }
+
+  auto& imageGeom = data.getDataRefAs<ImageGeom>(imageGeomPath);
+  SizeVec3 udims = imageGeom.getDimensions();
+  const auto imageGeomNumX = imageGeom.getNumXPoints();
+  const auto imageGeomNumY = imageGeom.getNumYPoints();
+  const auto imageGeomNumZ = imageGeom.getNumZPoints();
+
+  int64 dims[3] = {
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/FindNeighbors.cpp:166]:
```suggestion
  std::array<int64, 3> dims = {
```

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/FindNeighbors.cpp [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+
+  int64 neighpoints[6] = {0, 0, 0, 0, 0, 0};
+  neighpoints[0] = -dims[0] * dims[1];
+  neighpoints[1] = -dims[0];
+  neighpoints[2] = -1;
+  neighpoints[3] = 1;
+  neighpoints[4] = dims[0];
+  neighpoints[5] = dims[0] * dims[1];
+
+  int64 column = 0, row = 0, plane = 0;
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/FindNeighbors.cpp:180]:
```suggestion
  int64 column = 0;
  int64 row = 0;
  int64 plane = 0;
```

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]

**Review by imikejackson — DISMISSED** [review]

File: src/complex/Filter/Actions/CreateArrayAction.cpp [hunk 11]
```diff
@@ -12,7 +12,7 @@ namespace complex
 {
 CreateArrayAction::CreateArrayAction(DataType type, const std::vector<usize>& tDims, const std::vector<usize>& cDims, const DataPath& path)
 : IDataCreationAction(path)
-, m_Type(type)
+, m_Type(static_cast<DataType>(type))
```
Comment (JDuffeyBQ) [inline src/complex/Filter/Actions/CreateArrayAction.cpp:15]:
```suggestion
, m_Type(type)
```

File: src/complex/DataStructure/AbstractDataStore.hpp [hunk 12]
```diff
@@ -236,17 +272,25 @@ class AbstractDataStore : public IDataStore
...
     reference operator*() const
     {
-      return m_DataStore.get()[m_Index];
+      return (*m_DataStore)[m_Index];
     }
 
     bool operator==(const ConstIterator& rhs) const
     {
-      return (std::addressof(m_DataStore.get()) == std::addressof(rhs.m_DataStore.get())) && (m_Index == rhs.m_Index);
+      if(!isValid() && !rhs.isValid())
```
Comment (JDuffeyBQ) [inline src/complex/DataStructure/AbstractDataStore.hpp:289]: What if only one is invalid?

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]

**Review by JDuffeyBQ — APPROVED** [review]

**Pull request merged**